serialization/src/tagged.rs

```diff
@@ -0,0 +1,208 @@
+//! detect which variant are we dealing with.
+//!
+//! Here are the rules:
+//!
+//! * Only applicable to `enum`s
+//! * All variants must have data payload (at least a [Tag]), and the payload has to be [Tagged].
+//! * All `<_ as Tagged>::TAG` values must be different
+//!
+//! # Example
+//!
+//! Example demonstrates block header versioning:
+//!
+//! ```
+//! # use serialization::{*, tagged::*};
+//! // An initial header version with PoW
+//! #[derive(Tagged, Encode, Decode, Clone)]
+//! struct HeaderV1 {
+//!     version: Tag<1>,
+//!     prev_hash: [u8; 32],
+//!     tx_root: [u8; 32],
+//!     nonce: u128,
+//! }
+//!
+//! // A new version, switching to PoS and added timestamp
+//! #[derive(Tagged, Encode, Decode, Clone)]
+//! struct HeaderV2 {
+//!     version: Tag<2>,
+//!     prev_hash: [u8; 32],
+//!     tx_root: [u8; 32],
+//!     time: u32,
+//!     signature: Vec<u8>,
+//! }
+//!
+//! // A top-level header type
+//! #[derive(DirectEncode, DirectDecode)]
+//! enum Header {
+//!     V1(HeaderV1),
+//!     v2(HeaderV2),
+//! }
+//!
+//! // Create a PoW header
+//! let header_v1 = HeaderV1 {
+//!     version: Tag,
+//!     prev_hash: [42; 32],
+//!     tx_root: [43; 32],
+//!     nonce: 177_984_342_498_312,
+//! };
+//! let header = Header::V1(header_v1.clone());
+//!
+//! // Serializing `Header` and `HeaderV1` gives the same encoding.
+//! assert_eq!(header.encode(), header_v1.encode());
+//! ```
+//!
+//! ## Illegal usage
+//!
+//! It is mandatory to specify the initial byte value, the following is not legal:
+//!
+//! ```compile_fail
+//! # use serialization::tagged::*;
+//! #[derive(Tagged)]
+//! enum Example { X }
+//! ```
+//!
+//! While this is:
+//!
+//! ```
+//! # use serialization::tagged::*;
+//! #[derive(Tagged)]
+//! enum Example { X = 5 }
+//! ```
+//!
+//! Tagged enums cannot have multiple variants:
+//! ```compile_fail
+//! # use serialization::tagged::*;
+//! #[derive(Tagged)]
+//! enum Example { X = 1, Y = 2 }
+//! ```
+//!
+//! Directly encoded enums have to have distinct tag values:
+//! ```compile_fail
+//! # use serialization::tagged::*;
+//! #[derive(DirectDecode)]
+//! enum Example { X(Tag<5>), Y(Tag<5>) }
+//! ```
+
+/// Derive the [Tagged] trait
+///
+/// * An `enum` with just one variant always starts with the variant index and can be made [Tagged].
+///   The `#[codec(index = X)]` or enum discriminant has to be provided explicitly.
+/// * A struct where the first element implements [Tagged] can be made Tagged.
+pub use serialization_derive::Tagged;
+
+/// Derive [Encode] that does not add the discriminant byte
+pub use serialization_derive::DirectEncode;
+
+/// Derive [Decode] for decoding without the initial discriminant byte
+pub use serialization_derive::DirectDecode;
+
+use crate::{Decode, Encode, Input};
+
+/// Types whose encoding starts with a fixed byte (tag)
+pub trait Tagged {
+    /// Initial byte in encoding of `Self`
+    const TAG: u8;
```

Comment:
Can the tag be of a custom type? Nothing crazy, but other stack-based types, like u32 or whatever other size we need?

Reply:
I don't currently see a simple way to introduce the option to make this type customisable. It's takes a part in a bunch of const generics. Even if we figure out how to do it, I don't think it would be worth the extra complexity.

We could pick a different fixed type but I decided for `u8` for the following reasons:
* Compatibility with how are enums represented in SCALE (e.g. porting the existing Transaction enum to this would not change the encoding)
* I think it is unlikely that more than 256 versions/variants will be needed
* Even if over 256 variants are needed for some stuff
  * that is easily worked around by having nested versions when the need arises
  * the types where over 256 variants are not needed would have a needlessly wide tag representation